docs: rewrite documents docs (#16298)
Automattic/mongoose: docs/documents.md

You can create a new document using `new User()` or `await User.create()`; or load an existing document from MongoDB using queries like `findOne()`.

```javascript
// Create a new document in memory
const doc = new User({ name: 'John Smith' });
// Persist the document to MongoDB:
// the document is not persisted to MongoDB until you call `save()`.
await doc.save();

// Load an existing document
const existingDoc = await User.findOne({ name: 'John Smith' });

existingDoc instanceof User; // true
existingDoc instanceof mongoose.Model; // true
existingDoc instanceof mongoose.Document; // true
```

## Hydrated Documents vs Lean Documents

When you create a new document or load a document using a query, Mongoose returns a **hydrated document**.
Mongoose documents are not [plain-old JavaScript objects](https://masteringjs.io/tutorials/fundamentals/pojo): they are an instance of Mongoose's `Document` class.
In particular, Mongoose documents store internal state for:

* change tracking
* validation
* middleware
* document methods
* `save()`
* getters, setters, and virtuals

> That means some JavaScript object operations behave differently with documents.

Reviewer: Might be helpful to list what exact operations are and are not supported.

```javascript
// Does **not** throw
doc.age = 'not a number';

// Throws a ValidationError with a CastError for path `age`
await doc.validate();
```

This behavior allows Mongoose to collect multiple validation and casting errors together.
Also, if you set a value multiple times, the last value wins, so if you overwrite an invalid value then validation will succeed.

```javascript
doc.age = 'not a number';
doc.age = 42;

// Validation succeeds
await doc.validate();
```

**Validation** is a separate step that runs when you call the document's `validate()` method.
The `save()` method calls `validate()` internally, so `save()` also triggers validation.
Validation checks whether the document satisfies schema rules, which includes checking for cast errors, but also:

* [required](https://mongoosejs.com/docs/api/schematype.html#SchemaType.prototype.required())
* [min](https://mongoosejs.com/docs/api/schemanumber.html#SchemaNumber.prototype.min())
* [max](https://mongoosejs.com/docs/api/schemanumber.html#SchemaNumber.prototype.max())
* [enum](https://mongoosejs.com/docs/api/schemastring.html#SchemaString.prototype.enum())

> * custom validators

Reviewer: why is this not elaborated on?

# Documents

Mongoose documents are JavaScript objects backed by MongoDB data.
They are instances of Mongoose's [Document](https://mongoosejs.com/docs/api/document.html) class with built-in support for change tracking, casting, validation, middleware, and persistence.

<ul class="toc">
  <li><a href="#what-is-a-document">What is a Document</a></li>
  <li><a href="#hydrated-documents-vs-lean-documents">Hydrated Documents vs Lean Documents</a></li>
  <li><a href="#updating-using-save">Updating Using <code>save()</code></a></li>
  <li><a href="#setting-nested-properties">Setting Nested Properties</a></li>
  <li><a href="#casting-and-validation">Casting and Validation</a></li>
  <li><a href="#required-properties">Required Properties</a></li>
  <li><a href="#middleware">Middleware</a></li>
</ul>

## What is a Document?

> A Mongoose [document](api/document.html#Document) is an instance of a [Model](api/model.html#Model).

Reviewer 2: ```suggestion
A Mongoose [document](https://mongoosejs.com/docs/api/document.html#Document) is an instance of a [Model](https://mongoosejs.com/docs/api/model.html#Model).
```

Links should be changed to be consistently relative (`/api/...`) or always start with the root (`https://mongoosejs.com/...`)

When you instantiate a model, you create a new document.

```javascript
const User = mongoose.model('User', new Schema({
  name: String
}));

const doc = new User({ name: 'John Smith' });

doc instanceof User; // true
doc instanceof mongoose.Model; // true
doc instanceof mongoose.Document; // true
```

Mongoose documents represent individual documents stored in a MongoDB collection.
For example, documents created from the `User` model in the above example are stored in the `users` collection by default.

You can create a new document using `new User()` or `await User.create()`; or load an existing document from MongoDB using queries like `findOne()`.

```javascript

> const doc = new User({ name: 'John Smith' });
> // Persist the document to MongoDB. Calling `new User()` creates a new document in memory:
> // the document is not persisted to MongoDB until you call `save()`.
> await doc.save();

Reviewer 3: I think adding a separate comment first before `new User` reads a bit easier.
```suggestion
// Create a new document in memory
const doc = new User({ name: 'John Smith' });
// Persist the document to MongoDB:
// the document is not persisted to MongoDB until you call `save()`.
await doc.save();
```

Also, if you set a value multiple times, the last value wins, so if you overwrite an invalid value then validation will succeed.

```javascript
doc.age = 'not a number';
doc.age = 42;

// Validation succeeds
await doc.validate();
```

**Validation** is a separate step that runs when you call the document's `validate()` method.
The `save()` method calls `validate()` internally, so `save()` also triggers validation.
Validation checks whether the document satisfies schema rules, which includes checking for cast errors, but also:

* [required](https://mongoosejs.com/docs/api/schematype.html#SchemaType.prototype.required())
* [min](https://mongoosejs.com/docs/api/schemanumber.html#SchemaNumber.prototype.min())
* [max](https://mongoosejs.com/docs/api/schemanumber.html#SchemaNumber.prototype.max())
* [enum](https://mongoosejs.com/docs/api/schemastring.html#SchemaString.prototype.enum())
* custom validators

If validation fails, Mongoose does **not** save the document.

```javascript
const schema = new Schema({
  age: {
    type: Number,
    min: 0
  }
});

const Person = mongoose.model('Person', schema);

const doc = new Person({ age: -1 });

await doc.validate();

> // Path `age` (-1) is less than minimum allowed value (0)

Reviewer 3: Moved comment on top, and prefixed with `Throws an error:`
```suggestion
// Throws an error "Path `age` (-1) is less than minimum allowed value (0)"
await doc.validate();
```

```

## Required Properties

The most commonly used validator in Mongoose is `required`.
If a required path is missing when you validate or save a document, Mongoose throws a validation error.

```javascript
const schema = new Schema({
  name: {
    type: String,
    required: true
  }
});

const User = mongoose.model('User', schema);

const doc = new User();

// Throws an error "Path `name` is required"
await doc.validate();
```

For most schema types, any value that is not `null` or `undefined` will pass the required validator.
The exceptions are strings and buffers: empty string and empty buffer cause a `ValidationError` if `required` is set.

```javascript
const doc = new User({ name: '' });

// Throws an error "Path `name` is required"
await doc.validate();
```

Note that empty arrays do **not** cause a `ValidationError` if the array is `required`.

## Middleware

Document [middleware](https://mongoosejs.com/docs/middleware.html) lets you run code during key parts of a document's lifecycle.
The most common document middleware hooks are for `validate()` and `save()`.

At a high level, saving a document looks like this:

> <img src="https://res.cloudinary.com/drfhhq8wu/image/upload/v1779818614/bf1e3e29-7554-4ff9-9564-f631ed3eef98_ac2cb2.png" alt="save document lifecycle diagram">

Reviewer 2: The inline-codeblocks in this image still show backticks, is this intentional?

Author: Not intentional, fixed

<img width="1000" height="667" alt="image" src="https://github.com/user-attachments/assets/a606d870-b388-4077-a8df-0bf88b90d76f" />

```javascript
const userSchema = new Schema({
  name: String,
  normalizedName: String
});

userSchema.pre('validate', function() {
  if (this.name != null) {
    this.normalizedName = this.name.trim().toLowerCase();
  }
});

userSchema.pre('save', function() {
  console.log('Saving user:', this.name);
});

const User = mongoose.model('User', userSchema);

const user = new User({ name: '  JOHN SMITH  ' });
await user.save();
user.normalizedName; // 'john smith'
```

Document middleware is a good fit for logic that is closely tied to the document itself, such as:

* normalizing values before validation
* deriving one path from another
* enforcing document-level invariants
* logging or auditing document saves
* updating timestamps or metadata

However, use middleware sparingly.
Middleware is usually not a good place for complex application logic.
In particular, we recommend avoiding code in middleware that:

* makes network calls to unrelated services
* depends heavily on request-specific context
* performs expensive work that should be explicit

Also note that document middleware only runs for document operations.

> Query updates like `updateOne()` and `findOneAndUpdate()` do **not** run `save()` middleware.

Reviewer 2: Maybe a link to all the event names and where they get triggered would be good here.

Author: What do you mean by this - list all middleware functions that developers can hook into? I'm not sure the documents documentation is the right place for that.

Reviewer 2: Now that you mention it, yeah it likely is not the correct place. Also the link i meant was to `https://mongoosejs.com/docs/middleware.html`, which is already covered above, just below the section header.
